### PR TITLE
table shard bootstrap process (part 1)

### PR DESCRIPTION
**Pull request by jshencode**
diff description:
add table shard bootstrap process:
   1. fetch metadata from peer
   2. set metadata to local metastore
   3. fetch raw vector parties from peer and write to diskstore

mainly affected package/files:
1. datanode/client/types.go (interface for peerSource)
2. datanode/generated/proto/rpc (grpc client mocks for testing)
3. memstore/bootstrap.go (new file to hold all bootstrap related code at table shard level)
4. memstore/bootstrap_test.go (bootstrap tests)
5. cluster/  (only movement of controller/cluster package to cluster package and rename)

Follow up diffs:
1. implement peerSource which serves as source for managing connections to peers
2. add retry logic for failed vector parties

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -31,3 +31,4 @@ include/
 rdkafka-prefix/
 rmm-prefix/
 testing/data/
+go.sum
```
Comment (voyager-dw) [inline .gitignore]: move to .gitignore?

File: datanode/client/types.go [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+// ConnectionPool
+type ConnectionPool interface {
+	// Open starts the connection pool connecting and health checking.
+	Open()
+
+	// ConnectionCount gets the current open connection count.
+	ConnectionCount() int
+
+	// NextClient gets the next client for use by the connection pool.
+	NextClient() (rpc.PeerDataNodeClient, error)
```
Comment (voyager-dw) [inline datanode/client/types.go]: borrow/return function name?

File: .gitignore (hunk 1 above)
Comment (jshencode) [reply .gitignore]: this is .gitignore

File: cluster/store/etcd.go [hunk 3]
```diff
@@ -11,7 +11,7 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
-package cluster
+package store
```
Comment (voyager-dw) [inline cluster/store/etcd.go]: kvstore? we have many different stores now

File: .gitignore (hunk 1 above)
Comment (shz117) [reply .gitignore]: it's already on master, maybe rebase

File: memstore/bootstrap.go [hunk 4]
```diff
@@ -0,0 +1,479 @@
...
+		return nil, utils.StackError(err, "failed to create keep alive stream")
+	}
+
+	// send loop
+	go func(stream rpc.PeerDataNode_KeepAliveClient) {
+		for {
+			ticker := time.NewTicker(time.Duration(atomic.LoadInt64(&ttl) / 2))
+			select {
+			case <-ticker.C:
+				err = xretry.NewRetrier(xretry.NewOptions()).Attempt(func() error {
```
Comment (voyager-dw) [inline memstore/bootstrap.go]: nice, looks like i can borrow the retry logic

File: memstore/bootstrap.go [hunk 5]
```diff
@@ -0,0 +1,479 @@
...
+		} else {
+			shard.bootstrapState = bootstrap.BootstrapNotStarted
+		}
+		shard.bootstrapLock.Unlock()
+	}()
+
+	// find peer node for copy data
+	peerNode := shard.findBootstrapSource(origin, topo, topoState)
+	var dataStreamErr error
+	borrowErr := peerSource.BorrowConnection(peerNode.ID(), func(nodeClient rpc.PeerDataNodeClient) {
```
Comment (voyager-dw) [inline memstore/bootstrap.go]: so looks like one shard copy will use one connection, possible to distribute to multiple connection?

Comment (jshencode) [reply memstore/bootstrap.go]: grpc connection is ok to be shared and concurrently accessed. actually the number of streams is recommended to not be less than 100.
actually we will have a tighter limit on the number of concurrent raw data fetch streams (per vector party), enforced by the worker pool capacity.

File: memstore/bootstrap.go [hunk 6]
```diff
@@ -0,0 +1,479 @@
...
+		return err
+	}
+
+	// 2. set metadata and trigger recovery
+	if err := shard.setTableShardMetadata(tableShardMeta); err != nil {
+		return err
+	}
+
+	// 3. fetch raw vps
+	workerPool := xsync.NewWorkerPool(int(math.Ceil(float64(runtime.NumCPU()) / 2)))
```
[the remark this answers is not in the document]
Comment (jshencode) [reply memstore/bootstrap.go]: we should move these two into settings under bootstrap, concurrent_table_shard and concurrent_streams_per_table_shard

File: datanode/datanode.go [hunk 7]
```diff
@@ -210,12 +212,14 @@ func (d *dataNode) Bootstrap(topo topology.Topology) error {
...
 		workers.Go(func() {
-			err := tableShard.Bootstrap(d.host, topo, topoStateSnapshot)
+			err := tableShard.Bootstrap(d.peers, d.host, topo, topoStateSnapshot)
 
 			mutex.Lock()
 			multiErr = multiErr.Add(err)
 			mutex.Unlock()
 
+			// unpin table shard after use
+			tableShard.Users.Done()
```
[the remark this answers is not in the document]
Comment (jshencode) [reply datanode/datanode.go]: yeah. good catch.

**Pull request merged**